alonzo/impl/src/Cardano/Ledger/Alonzo/Data.hs

```diff
@@ -166,40 +182,105 @@ deriving via InspectHeapNamed "AuxiliaryDataRaw" (AuxiliaryDataRaw era) instance
 instance
   ( Typeable era,
     Ord (Core.Script era),
-    ToCBOR (Core.Script era)
+    Core.Script era ~ Script era,
+    ToCBOR (Core.Script era),
+    Typeable (Crypto era)
   ) =>
   ToCBOR (AuxiliaryDataRaw era)
   where
   toCBOR (AuxiliaryDataRaw s d m) =
     encode (encodeRaw s d m)
 
 encodeRaw ::
-  (ToCBOR (Core.Script era), Typeable era) =>
+  ( ToCBOR (Core.Script era),
+    Core.Script era ~ Script era,
+    Typeable (Crypto era),
+    Typeable era
+  ) =>
   Map Word64 Metadatum ->
   StrictSeq (Core.Script era) ->
   Set (Data era) ->
-  Encode ('Closed 'Dense) (AuxiliaryDataRaw era)
-encodeRaw m s d =
-  ( Rec AuxiliaryDataRaw
-      !> mapEncode m
-      !> E encodeFoldable s
-      !> setEncode d
+  Encode ('Closed 'Sparse) (AuxiliaryDataRaw era)
+encodeRaw metadata scripts dats =
+  ( Tag 259 $
+      Keyed
+        (\m tss pss d -> AuxiliaryDataRaw m (StrictSeq.fromList $ tss <> pss) d)
+        !> Omit null (Key 0 $ mapEncode metadata)
+        !> Omit null (Key 1 $ E (encodeFoldable . mapMaybe getTimelock) timelocks)
+        !> Omit null (Key 2 $ E (encodeFoldable . mapMaybe getPlutus) plutusScripts)
+        !> Omit null (Key 3 $ setEncode dats)
   )
+  where
+    getTimelock (TimelockScript x) = Just x
+    getTimelock _ = Nothing
+    getPlutus (PlutusScript x) = Just x
+    getPlutus _ = Nothing
+    (plutusScripts, timelocks) =
+      List.partition
+        isPlutusScript
+        (Foldable.toList scripts)
```

Comment:
One could fold the stripping of the Alonzo construtors (TimelockScript and Plutuscript) and the partioning into a single function. But then you'd have to use a fold. :-)

splitAndStrip xs = foldl accum ([],[]) xs where
  accum (ts,ps) (PlutusScript p) = (ts,p:ps)
  accum (ts,ps) (TimelockScript t) = (t:ts,ps)